Improve the --help menu and add the --version and -v flags to the command-line interface

gkapfham opened the pull request:
<!-- TODO: Replace the title below -->
# Improve the help menu and command-line interface

## Description

Currently it is not possible to determine what version of GatorGrade you have installed through the use of either `--help` or the `--version` flags. This means that the only way you can know the version of `gatorgrade` that you are using is to carefully observe this information when you install it with `pipx` or to, alternatively, uninstall and then reinstall `gatorgrade`.

This is what the output now looks like when you use `--help`:

![image](https://user-images.githubusercontent.com/926029/189463505-5bb14088-7fb0-4df9-bbda-a9f08f0639ff.png)

This is what the output now looks like when you use `--version`:

![image](https://user-images.githubusercontent.com/926029/189463625-3459f51c-5a63-4304-b3f0-ff520c2cc061.png)

If you want to install and try out this version you can use this command:

```
pipx install git+https://github.com/GatorEducator/gatorgrade.git@feat/add-version
```

## Linked Issues

None.

## Type of Change

<!-- TODO: Fill in the brackets with an `x` next to all types that apply to the proposed changes -->
- [X] Feature
- [ ] Bug fix
- [ ] Documentation

## Contributors

<!-- TODO: Add your GitHub username below and the GitHub usernames of all other contributors to the proposed changes -->
- @gkapfham

## Reminder

 - [X] All GitHub Actions should be in a passing state before any pull request is merged.
 - [X] All PRs must be reviewed by at least one team member and one member of the Integration team!
 - [X] Any issues this PR closes are tagged in the description!

ullrichd21 committed:
Add JSON output with "--report" or "-r" flags.

ullrichd21 committed:
Fix test to support returned tuple.

ullrichd21 committed:
Fix codestyle/linting.

ullrichd21 committed:
Fix codestyle/linting, add encoding type.

ullrichd21 committed:
Fix codestyle/linting.

gkapfham committed:
feat: Add draft of new help message with version info in main.

gkapfham committed:
feat: Add version module with get_gatorgrade_version and
get_gatorgrader_version.

gkapfham committed:
fix: Refine the phrase in the epilog_message in main.

gkapfham committed:
refactor: Reorganize gatorgrade function in main to handle version
parameter.

gkapfham committed:
fix: Delete the extra help_message not needed in main module.

gkapfham committed:
feat: Add get_project_version and refactor version functions.

gkapfham committed:
refactor: Use improved markdown version strings in the main module.

gkapfham committed:
refactor: Remove a not-needed and commented-out global message in main.

gkapfham committed:
feat: Add constants and all projects in PROJECTS list in version.

gkapfham committed:
refactor: Further improve help and epilog messages in main.

gkapfham committed:
fix: Remove spaces in a multi-line string in main.

gkapfham committed:
fix: Remove spaces in a multi-line string in main.

gkapfham committed:
refactor: Add version module in the util package.

gkapfham committed:
refactor: Rewrite import for version to be in gatorgrade.util in main.

gkapfham committed:
refactor: Delete the version module in top-level package.

gkapfham committed:
coms: Add two and update one comment in main module.

gkapfham committed:
fix: Delete conditional logic on help that was not needed in main.

gkapfham committed:
feat: Add draft of the --version output in main.

gkapfham committed:
feat: Add versions module in util to not conflict with --version in
main.

gkapfham committed:
refactor: Remove version module that conflicts with --version in main.

gkapfham committed:
feat: Improve the help_message_markdown in main.

gkapfham committed:
coms: Improve a comment in gatorgrade function in main.

gkapfham committed:
refactor: Make the variable GATORGRADE_EMOJI in main module constants.

gkapfham committed:
feat: Improve main so that github_message is generated.

gkapfham committed:
style: Reformat the source code in versions.

gkapfham committed:
feat: Add the github module in util with get_github_projects function.

gkapfham committed:
refactor: Use the GATORGRADE_EMOJI_RICH constant in main.

gkapfham committed:
fix: Delete not-needed variables for GitHub projects in main.

gkapfham committed:
refactor: Use the RICH_MARKUP_MODE_DEFAULT constant in main.

gkapfham committed:
chore: Increase minimal version of Python to 3.8 in pyproject.toml.

gkapfham committed:
chore: Use Python 3.8 and 3.10 on ubuntu-latest and 3.9 on macos-latest
and windows-latest in main.yml.

gkapfham committed:
chore: Bump the poetry-version to 1.2 in the main.yml file.

gkapfham committed:
chore: Update poetry.lock file to reflect move to Python 3.8 minimum.

gkapfham committed:
fix: Revert/increment poetry-version in main.yml to 1.1.14 as 1.2 fails.

gkapfham committed:
Merge branch 'main' into feat/add-version

gkapfham committed:
chore: Specify Python 3.8 as the minimum version in publish.yml.

burgess01 reviewed (approved):
Looks great! I just had one question. I saw in the versions.py file that you had functions specifically for getting the versions of gatorgrade and gatorgrader, and that these functions aren't currently in use. Especially since these don't have formatting like the others do, are you planning on using these in the future or are they just if-needed? Thanks!

gkapfham commented:
Hello @burgess01, thanks for pointing out this oversight!

I'm pretty sure that those functions are no longer needed and should be removed from the code base before merging this PR.

I will review everything again on Monday and most likely remove those functions and make additional tweaks as needed.

gkapfham committed:
Merge branch 'main' into feat/add-version

gkapfham committed:
coms: Add a comment to versions to explain PROJECTS variable.

gkapfham committed:
coms: Fix an incorrect docstring in the versions module.

gkapfham committed:
coms: Add comments to explain steps in the get_project_versions in versions.

gkapfham committed:
refactor: Delete two not-needed functions in the versions module.

gkapfham committed:
fix: Stop using an epilog in main and add more to version output.

gkapfham committed:
feat: Improve the output of --version in the main module.

gkapfham committed:
coms: Improve two comments in the main module.

gkapfham committed:
fix: Delete the GATORGRADE_EMOJI constant in main.

gkapfham committed:
refactor: Delete a not-needed variable in main.

gkapfham committed:
coms: Add comments to explain steps in get_github_projects in github.

gkapfham committed:
coms: Delete a not-needed comment for typer app in main.

gkapfham committed:
coms: Clarify the docstring comment in main.

gkapfham committed:
chore: Bump the semver in the pyproject.toml file.

Yanqiao4396 committed:
upgrade python version to unification

Michionlion committed:
Update GatorGrader version and regenerate poetry.lock with new versions

Michionlion committed:
Merge pull request #108 from GatorEducator/update-python-version

upgrade python version to unification

burgess01 committed:
Merge branch 'main' of github.com:GatorEducator/gatorgrade into feature/json-output

burgess01 committed:
feature: remove files no longer in main

burgess01 committed:
feature: make the report file param more robust

burgess01 committed:
feature: create funct to create checks json file

burgess01 committed:
feature: adjust tests to account for file param

burgess01 committed:
docs: add insights json file to test repo

burgess01 committed:
feature: update json to have more check info

burgess01 committed:
feature: incorporate md conversion from json

burgess01 committed:
feature: give more info about failing checks

burgess01 committed:
docs: correct md converter docstring

Yanqiao4396 committed:
Bug fix: delete one line in out

burgess01 committed:
feature: rework json and md funcs

burgess01 committed:
feature: rework json and md funcs

burgess01 committed:
fix: correct missing -- in tag catch

burgess01 committed:
docs: update test json and md files

burgess01 committed:
fix: rework icons to pass in windows

burgess01 committed:
fix: remove font to fix windows error

burgess01 committed:
fix: encode md file to fix windows error

burgess01 committed:
feature: make report a bool instead of Path

burgess01 committed:
feature: refactor functions to not use files

burgess01 committed:
feature: remove json and md files from test cases

burgess01 committed:
docs: correct function docstrings

burgess01 committed:
docs: add test cases for json and md files

burgess01 committed:
docs: add os import for test case file removal

burgess01 committed:
feature: add tuple based cli functionality

burgess01 committed:
feature: add function to interpret tuple cli

burgess01 committed:
fix: close files to avoid windows issues

burgess01 committed:
fix: encode file for windows compatibility

burgess01 committed:
fix: refactor md test case to work on all os

Yanqiao4396 committed:
Add env feature

Yanqiao4396 committed:
Add help codes

Yanqiao4396 committed:
Bug fix

burgess01 committed:
fix: change logic to pass test cases

burgess01 committed:
fix: correct black linting error in output.py

burgess01 committed:
feature: change any func call to all

Yanqiao4396 committed:
Add test for case of invalid path input

Yanqiao4396 committed:
Refactor json format

Yanqiao4396 committed:
Change related test

Yanqiao4396 committed:
Correct Lint

Yanqiao4396 committed:
fix quote sign to double for json generation

burgess01 committed:
feature: change format of md file output

burgess01 committed:
feature: edit formatting of md file failed checks

burgess01 committed:
feature: add param to both check classes

burgess01 committed:
feature: incorporate new param into class creation

burgess01 committed:
feature: add json param to result class

burgess01 committed:
feature: update output to work w new class struct

burgess01 committed:
feature: rework json/md funcs to use new classes

burgess01 committed:
feature: correct linting

Yanqiao4396 committed:
Add new test

burgess01 committed:
fix: reformat using black

burgess01 committed:
fix: correct json test case

Yanqiao4396 committed:
Fix: check box in failing section in md

Yanqiao4396 committed:
Fox: env variable create

Yanqiao4396 committed:
Fix: change back

Yanqiao4396 committed:
Fix: env variable creation in github action

Yanqiao4396 committed:
docs: add comment for --help

burgess01 committed:
feature: include project name in md output

burgess01 committed:
docs: update test case for md output change

burgess01 committed:
feature: readd previous changes

burgess01 committed:
feature: combine amt and % lines, fix tests

Yanqiao4396 committed:
bug: fix display diagnosis msg in terminal

Yanqiao4396 committed:
Feat: add diagnostic message in js and md output

To align with the new feature of diagnosis in the terminal
output, the json

Yanqiao4396 committed:
Factor: lint fix

burgess01 committed:
Merge pull request #111 from GatorEducator/feature/json-output

Create json and md checks output

burgess01 committed:
Merge pull request #113 from GatorEducator/bug/display-diagnosis

bug: fix display diagnosis msg in terminal

Yanqiao4396 committed:
Delete malicious link

Yanqiao4396 committed:
Add new hint

Yanqiao4396 committed:
Factor: Fix markdown lint

Yanqiao4396 committed:
Docs: remove meangingless hint

gkapfham commented:
Hi @burgess01 and @Yanqiao4396 --- this PR is now very out of date.

However, once we have repaired the build and made a release of your new feature(s), perhaps we can revisit this PR, patch it up, and then release it so that these flags are available?

Yanqiao4396 committed:
Merge pull request #115 from GatorEducator/Docs/Improve-README

Remove malicious link in README.md

Yanqiao4396 committed:
Bug: Upgrade poetry ver for http incompatibility

Yanqiao4396 committed:
Bug: pin urll3 <=2.0

Yanqiao4396 committed:
Bug: pin urllib3 lower than 2 by pip

Yanqiao4396 committed:
Bug: Upgrade Poetry version to newest

Try out newest poetry build to fix bug http requests
has no attribute strict

Yanqiao4396 committed:
Factor: Remove useless comment

Yanqiao4396 committed:
Bug: Remove urllab3 dep in pyproject.toml

Specifying urllab3 in pyproject.toml doesn't help with the Requests bug. Delete it

Yanqiao4396 committed:
Feat: Upgrade Poetry to 1.5.0

Upgrade Poetry in the workflow/main.yml to the newest version 1.5.0

Yanqiao4396 committed:
Merge pull request #117 from GatorEducator/HTTP-incompatibility

Fix Attribute Error: 'HTTPResponse' object has no attribute 'strict'

Yanqiao4396 committed:
Bug: upgrade setup-poetry

setup-poetry V7 fails for HTTP request 404

Yanqiao4396 committed:
Merge pull request #120 from GatorEducator/bug/poetry-build-fail

Bug/poetry-setup build fail

Yanqiao4396 committed:
Bug: mkdocs poetry update to v8

Yanqiao4396 committed:
Refac: distinguish json report from md report

json report and md report used to share
the same environment
variable. Set two different variables now as we may need both
of them in the future code

Yanqiao4396 committed:
Feat: add json report in the github env

add json report in the github env for the future data collection

Yanqiao4396 committed:
Merge branch 'main' into feat/add-env-to-github-env
main is important# the commit.

Yanqiao4396 committed:
update poetry build to v8 in publish.yml

Yanqiao4396 committed:
Refact: put file write code into a func

file write related code is used more than 3 times, put it into a
function to make the design dryer

Yanqiao4396 committed:
Add new test for the refactor

Yanqiao4396 committed:
Feat: always add json_report into github_env

Yanqiao4396 committed:
Feat: json dump replace dumps with test

dumps write all the content in the first line
of json file, which is bad. Replace it with dump to add indentation

Yanqiao4396 committed:
Test: add test for file write related code

Yanqiao4396 committed:
Merge pull request #121 from GatorEducator/bug/poetry-build-fail

Bug: mkdocs poetry update to v8

Yanqiao4396 committed:
Merge branch 'main' into feat/add-env-to-github-env

Yanqiao4396 committed:
Merge pull request #122 from GatorEducator/feat/add-env-to-github-env

Feature: Add json report to the GITHUB_ENV

Yanqiao4396 committed:
Bug: json dumps replaces str

used to use dict type and failed as dict is not
equal to json type. Now put json type into JSON_REPORT correctly

Yanqiao4396 committed:
Remove indent

Yanqiao4396 committed:
Merge pull request #123 from GatorEducator/bug/JSON_REPORT-dumps

Bug: json dumps replaces str

Yanqiao4396 committed:
Feat: add path in json_inf of CheckResult

It helps with identifying the file, path a command or
check belongs to in the report

Yanqiao4396 committed:
Test: customized json inf will be stored in report

Yanqiao4396 committed:
Bug: all front slash replacing back slash

Yanqiao4396 committed:
Bug: not have path attribute if path value is none

Yanqiao4396 committed:
Merge pull request #124 from GatorEducator/feat/add-path-to-report

Add Path Attribute to the CheckResult and Test Customized Insight Key Value Pair

Yanqiao4396 committed:
Feat: add current time to report

Yanqiao4396 committed:
Merge pull request #125 from GatorEducator/feat/time-in-report

Feat: add current time to json report

gkapfham committed:
chore: Start to use python-version 3.10 in publish.yml.

gkapfham committed:
chore: Improve a name of a step in publish.yml.

gkapfham committed:
chore: Add the poetry-version attribute to 1.5.0 in publish.yml

gkapfham committed:
chore: Adopt a different approach for installing poetry in publish.yml.

rebekahrudd committed:
feat: added command printing code in output.py

rebekahrudd committed:
fix: deleted an extra variable assignment in output.py

rebekahrudd committed:
fix: deleted extra comments in output.py

gkapfham committed:
refactor: Make the diagnostics message for failing command shorter in gatorgrade/output/output.py.

gkapfham committed:
chore: Update the poetry.lock file with new dependencies.

gkapfham committed:
chore: Bump the semver (it was 0.0.0, starting to use it again) in the pyproject.toml file.

gkapfham committed:
Merge branch 'feat/add-failing-command-diag'

rebekahrudd committed:
fix: implementation of output code and added feature in output.py

rebekahrudd committed:
fix: command output added for GatorGraderCheck in output.py in fork

rebekahrudd committed:
fix: made failed_results subscriptable again in output.py

rebekahrudd committed:
deleted: print statements in test_main.py and test_output.py

rebekahrudd committed:
fix: reformatted output.py, test_output.py, and test_main.py

rebekahrudd committed:
fix: import formatting in test_output.py

rebekahrudd committed:
fix: update the over general exceptions to pass linting checks in .pylintrc

rebekahrudd committed:
style: added an encoding type to the open file function in output.py

gkapfham committed:
chore: Add the poetry.toml file to the .gitignore file.

rebekahrudd committed:
chore: add a newline after command output in output.py

rebekahrudd committed:
delete: an unecessary comment in test_main.py

gkapfham committed:
debug: Comment out code for testing purposes in gatorgrade/output/output.py.

gkapfham committed:
Merge branch 'fix/report_bug' of github.com:GatorEducator/gatorgrade into fix/report_bug

gkapfham committed:
debug: Uncomment lines in the gatorgrade/output/output.py file for testing purposes.

rebekahrudd committed:
fix: ran poetry linter to fix formatting in output.py

gkapfham committed:
refactor: Remove the console.print() statements not needed in gatorgrade/output/output.py.

gkapfham committed:
Merge branch 'fix/report_bug' of github.com:GatorEducator/gatorgrade into fix/report_bug

gkapfham committed:
style: Remove the inconsistent and not-needed blank lines in gatorgrade/output/output.py.

gkapfham committed:
style: Remove a not-needed import and blank lines and reorganize imports in gatorgrade/output/output.py.

gkapfham committed:
style: Remove not-needed blank lines in the gatorgrade/output/output.py file.

gkapfham committed:
chore: Bump the semver to 0.7.0 in the pyproject.toml file and upgrade isort.

gkapfham committed:
style: Resort the imports according to isort as a stopgap measure for gatorgrade/output/output.py.

gkapfham committed:
chore: Update the poetry.lock file with new dependencies.

gkapfham committed:
refactor: Try a new approach to saving the failing commands in gatorgrade/output/output.py.

gkapfham committed:
refactor: Add a run_command as an null string as a field in gatorgrade/output/check_result.py and then add a __repr__ for debugging purposes.

gkapfham committed:
refactor: Comment out not-needed code and repair linting mistakes in gatorgrade/output/output.py.

gkapfham committed:
coms: Add more explanatory comments to gatorgrade/output/output.py.

gkapfham committed:
style: Remove the not-needed blank lines in the gatorgrade/output/output.py.

gkapfham committed:
style: Reformat all source code in gatorgrade/generate/generate.py with ruff.

gkapfham committed:
style: Reformat the gatorgrade/input/checks.py file to address ruff rules.

gkapfham committed:
fix: Delete variable(s) that are not used in the gatorgrade/output/output.py file.

gkapfham committed:
chore: Add new task and project details and dependencies to the pyproject.toml file.

gkapfham committed:
chore: Add the poetry.lock file with the new dependencies.

gkapfham committed:
fix: Remove an incorrect type annotation for a return of a function in gatorgrade/input/in_file_path.py.

gkapfham committed:
chore: Add type ignore pragmas to the gatorgrade/input/checks.py file.

gkapfham committed:
chore: Add type ignore pragmas to the gatorgrade/output/output.py file.

gkapfham committed:
fix: Use the Union type as part of an annotation to the gatorgrade/output/check_result.py file.

gkapfham committed:
fix: Add the correct type annotation for a list in the gatorgrade/input/command_line_generator.py file.

gkapfham committed:
chore: Delete the use of symbex in the pyproject.toml file because not all functions have type annotations.

gkapfham committed:
test: Delete flaky tests in tests/output/test_output.py that do not have correct character encoding.

gkapfham committed:
test: Delete a not-needed capsys in the tests/output/test_output.py file.

gkapfham committed:
chore: Add mypy to dev dependencies group in pyproject.toml and update poetry.lock.

gkapfham committed:
chore: Add mypy to and types-pyyaml to pyproject.toml and improve formatting.

gkapfham committed:
chore: Update the poetry.lock file with new dependencies.

gkapfham committed:
chore: Check the matrix.os to prevent tests being run on Windows in .github/workflows/main.yml.

gkapfham committed:
refactor: Improve the implementation of a function in gatorgrade/output/output.py so it only writes to GITHUB_ENV when it is set.

gkapfham committed:
coms: Add more comments to the gatorgrade/output/output.py file.

gkapfham committed:
refactor: Do not attempt to save to GITHUB_STEP_SUMMARY if it does not exist in gatorgrade/output/output.py.

gkapfham committed:
style: Reformat source code to meet ruff standards in gatorgrade/output/output.py.

gkapfham committed:
chore: Add @rebekahrudd as a contributor in the pyproject.toml file.

gkapfham committed:
Merge branch 'fix/report_bug'

gkapfham committed:
chore: Delete the mkdocs.yml file as no longer building a web site and this GitHub Actions workflow is no longer needed.

rebekahrudd committed:
feat: defined command output in check_result.py

rebekahrudd committed:
fix: changed the color of the failing command output in check_result.py

rebekahrudd committed:
Merge branch 'feat/add-version' of github.com:GatorEducator/gatorgrade into feat/add-version

rebekahrudd committed:
feat: added version to the command line in main.py

rebekahrudd committed:
format: main.py file reformatted

rebekahrudd committed:
format: main.pr reformatted

rebekahrudd commented:
This branch has been updated to current version of `gatorgrade`

Terminal output of the command `gatorgrade --version`
![s-version](https://github.com/user-attachments/assets/6eb9f95f-a90b-4099-83e7-ed69e54b7696)

Terminal output of command `gatorgrade --help`
![s-help](https://github.com/user-attachments/assets/5434f4fb-3635-48d2-aa98-de53b2bfd9a0)

rebekahrudd reviewed (approved):
This updated works on Linux, though the `--help` command does not show the complete version output.

dyga01 reviewed (approved):
These changes appear to be working on Mac Os. LGTM!
